### PR TITLE
Restoration type component

### PR DESCRIPTION
**Pull request by JazzarKarim**
*Issue #:* /bcgov/entity#14797

*Description of changes:*
- Created Restoration Type Component:

https://user-images.githubusercontent.com/122301442/217377618-0deb8343-d068-477e-965f-5642a8db2fe6.mp4

- Created Restoration Type Summary Component:

  - Full Restoration but no relationship is selected:
full error

  - Full Restoration with relationships selected:
full good

  - Limited Restoration selected:
Limited good

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the bcrs-entities-create-ui license (Apache 2.0).

### CODE REVIEW

File: src/interfaces/store-interfaces/state-interface.ts
Comment (severinbeauvais) [inline src/interfaces/store-interfaces/state-interface.ts:72]: Can this be moved to the restoration state object?

Comment (JazzarKarim) [reply src/interfaces/store-interfaces/state-interface.ts:72]: I thought about it. Thing is, restoration schema doesn't have an isValid field.

Comment (severinbeauvais) [reply src/interfaces/store-interfaces/state-interface.ts:72]: There is an interface for the restoration state, which is different from the restoration JSON object.

File: src/enums/index.ts
Comment (severinbeauvais) [inline src/enums/index.ts:31]: Please sort this list. It's not a big deal for a small list like this, but for larger lists it saves times/errors when looking for something.

File: src/constants/constants.ts
Comment (severinbeauvais) [inline src/constants/constants.ts:5]: Are these used anywhere?

Comment (severinbeauvais) [reply src/constants/constants.ts:5]: Oh, found it.

File: src/enums/index.ts
Comment (JazzarKarim) [reply src/enums/index.ts:31]: Noted.

File: src/interfaces/store-interfaces/state-interface.ts
Comment (JazzarKarim) [reply src/interfaces/store-interfaces/state-interface.ts:72]: Oh wait, I think I have things mixed up a little. In that case, I think I can move it to the state object.

File: src/constants/constants.ts
Comment (severinbeauvais) [reply src/constants/constants.ts:5]: This use of constants is a bad pattern. In your case, please create a new enum (type) and use that. This will also make `selectRestorationType` nicely typed. (You can assign null as an initial "unknown" value if needed.)

File: src/views/Restoration/RestorationBusinessName.vue
Comment (severinbeauvais) [inline src/views/Restoration/RestorationBusinessName.vue:13]: Please restore this. I will be updating it in ticket 14783.

File: src/views/Restoration/RestorationBusinessName.vue
Comment (severinbeauvais) [inline src/views/Restoration/RestorationBusinessName.vue:12]: This v-show is suspicious.
1. I'm not sure why you need to check this here. Is entity type not set sometimes?
2. Why it is a v-show instead of a v-if? Do the component within need to be rendered when we don't know entity type?

File: src/views/Restoration/RestorationReviewConfirm.vue
Comment (severinbeauvais) [inline src/views/Restoration/RestorationReviewConfirm.vue:100]: Is this component supposed to be rendered on this page? I don't see it.

Comment (JazzarKarim) [reply src/views/Restoration/RestorationReviewConfirm.vue:100]: I just imported it but I didn't render it since it's going to be a part of a bigger component. The component is going to have other things like Approval type. So I left there to be used later.

File: tests/unit/RestorationType.spec.ts
Comment (severinbeauvais) [inline tests/unit/RestorationType.spec.ts:17]: Fix or delete please.

File: tests/unit/RestorationType.spec.ts
Comment (severinbeauvais) [inline tests/unit/RestorationType.spec.ts:19]: If this sub-component has appropriate tests in the shared repo, then you don't need to test it again here.

However, the component that hosts this sub-component should make sure the sub-component is displayed when needed, and also verify that the right thing is done with any events (or getters/computed variables).

File: tests/unit/RestorationTypeSummary.spec.ts
Comment (severinbeauvais) [inline tests/unit/RestorationTypeSummary.spec.ts:13]: This test suite should also verify that store properties display correctly, plus any other things that this component does.

File: tests/unit/RestorationType.spec.ts
Comment (JazzarKarim) [reply tests/unit/RestorationType.spec.ts:19]: Yes, I was confused what should I exactly test since I've done a lot of the testing in the sub-components (bcrs shared repo).

File: src/components/Restoration/RestorationType.vue
Comment (severinbeauvais) [inline src/components/Restoration/RestorationType.vue:136]: Labels typically already have a base style that makes them bold.

If you're only styling a single label then I recommend using the Vuetify helper classes to set font weight: https://vuetifyjs.com/en/styles/text-and-typography/

File: src/components/Restoration/RestorationTypeSummary.vue
Comment (severinbeauvais) [inline src/components/Restoration/RestorationTypeSummary.vue:15]: How wide is this column? Can you show screenshots on a wide screen and narrow screen?

File: src/components/Restoration/RestorationTypeSummary.vue
Comment (severinbeauvais) [inline src/components/Restoration/RestorationTypeSummary.vue:20]: Please use enum for value.

File: src/views/Restoration/RestorationReviewConfirm.vue
Comment (severinbeauvais) [reply src/views/Restoration/RestorationReviewConfirm.vue:100]: Isn't it part of your ticket to show restoration/approval type on the review page? If not, do we need another ticket?

File: tests/unit/RestorationType.spec.ts
Comment (severinbeauvais) [reply tests/unit/RestorationType.spec.ts:19]: Then you don't need to test the sub-component again.

But you should test the props you would send to it and how you handle the events you would get back from it -- we don't do this nearly as well as we should, but we should try :)

Think "tiers" -- THIS component should be tested here, but not its parent or children components.

File: src/views/Restoration/RestorationReviewConfirm.vue
Comment (JazzarKarim) [reply src/views/Restoration/RestorationReviewConfirm.vue:100]: Yes it is! Okay, I'll add it to the review page. 

summary

This is what's going to look like in the end. My summary component is the one in the middle. I'll add and keep it in the review page.

File: tests/unit/RestorationType.spec.ts
Comment (severinbeauvais) [reply tests/unit/RestorationType.spec.ts:19]: If extra unit tests will take more than 1 day, I'm OK with creating a new Tech Debt ticket to expand them later.

File: src/components/Restoration/RestorationTypeSummary.vue
Comment (JazzarKarim) [reply src/components/Restoration/RestorationTypeSummary.vue:15]: Wide:
wide

Narrow:
narrow

Comment (severinbeauvais) [reply src/components/Restoration/RestorationTypeSummary.vue:15]: Can you show a little more on the narrow screen, please? I want to see how the other sections display.

File: src/components/Restoration/RestorationType.vue
Comment (severinbeauvais) [inline src/components/Restoration/RestorationType.vue:78]: Instead of managing this property, why not use a computed value (getter) instead?
```
get isFull (): boolean {
   return (this.getRestoration.type === RestorationTypes.FULL)
}
```
And ditto for `isLimited`.

Let the framework do the reactivity for you.

File: src/components/Restoration/RestorationTypeSummary.vue
Comment (JazzarKarim) [reply src/components/Restoration/RestorationTypeSummary.vue:15]: localhost_8080_businesses_create_restoration-review-confirm_id=BC0871154 accountid=3040 (1)

File: src/components/Restoration/RestorationType.vue
Comment (severinbeauvais) [inline src/components/Restoration/RestorationType.vue:94]: Instead of this code, did you try initializing `selectRestorationType` from the store value? I think the property initialization is done when this component is created (not mounted), but it may work.
```
protected selectRestorationType = this.getRestoration.type
```

File: src/components/Restoration/RestorationType.vue
Comment (severinbeauvais) [inline src/components/Restoration/RestorationType.vue:136]: You don't have to, but you can, delete this style block, since it's not being used.

File: src/components/Restoration/RestorationTypeSummary.vue
Comment (severinbeauvais) [inline src/components/Restoration/RestorationTypeSummary.vue:55]: We don't have an official code style (other than what our linter enforces), but typically we don't have blank lines at the top or bottom of a block.

File: src/store/mutations/mutations-model.ts
Comment (severinbeauvais) [inline src/store/mutations/mutations-model.ts:421]: There's a mismatch between the mutation name and the thing you're modifying. Please fix.

File: src/interfaces/store-interfaces/state-interfaces/restoration-state-interface.ts
Comment (severinbeauvais) [inline src/interfaces/store-interfaces/state-interfaces/restoration-state-interface.ts:15]: Sorry, I just noticed this now, but how about renaming this flag to be consistent with the other validity flags in this object?

File: src/views/Restoration/RestorationBusinessName.vue
Comment (severinbeauvais) [inline src/views/Restoration/RestorationBusinessName.vue:175]: Again, our non-official code style is to leave whitespace between separate blocks. Please restore it.

File: src/components/Restoration/RestorationTypeSummary.vue
Comment (severinbeauvais) [reply src/components/Restoration/RestorationTypeSummary.vue:15]: Even narrower!

What I want to check is that keys and values stack correctly on mobile, ie:
image

Comment (JazzarKarim) [reply src/components/Restoration/RestorationTypeSummary.vue:15]: localhost_8080_businesses_create_restoration-review-confirm_id=BC0871154 accountid=3040 (2)

Comment (severinbeauvais) [reply src/components/Restoration/RestorationTypeSummary.vue:15]: OK, nice!

File: src/components/Restoration/RestorationTypeSummary.vue
Comment (severinbeauvais) [inline src/components/Restoration/RestorationTypeSummary.vue:15]: There's some unwanted leading whitespace here.

Compare this with https://github.com/bcgov/business-create-ui/blob/bf73ca09f6f81fe527689ac2b1c43ed3e1df8626/src/components/Incorporation/SummaryDefineCompany.vue#L21

File: src/components/Restoration/RestorationType.vue
Comment (JazzarKarim) [reply src/components/Restoration/RestorationType.vue:94]: This won't work! I tried it out, `this.getRestoration.type` is returning undefined.

Comment (severinbeauvais) [reply src/components/Restoration/RestorationType.vue:94]: It might return null since that's what the value in the store is initialized to.

Anyway, this means this component is created before the restoration is fetched -- not surprising since the fetch is async (network call). I was hoping maybe this component wasn't rendered right away.

Side note: what would be the the harm in initializing `selectRestorationType` as null?

File: src/components/Restoration/RestorationType.vue
Comment (JazzarKarim) [reply src/components/Restoration/RestorationType.vue:78]: These two properties are used to determine which panel to expand based on radio button selection. Are you sure having computed value getters would help?

File: src/components/Restoration/RestorationType.vue
Comment (JazzarKarim) [reply src/components/Restoration/RestorationType.vue:94]: Yes makes sense. There's no harm, I could do that. This is because I'm setting it again based on what was previously selected in the `mounted` block.

File: src/components/Restoration/RestorationType.vue
Comment (severinbeauvais) [reply src/components/Restoration/RestorationType.vue:78]: I'm 99% sure. Getters are reactive -- they update depending on the properties they look at. And the template is also reactive, so I think your v-cards will conditionally display just fine.

File: src/components/Restoration/RestorationType.vue
Comment (severinbeauvais) [reply src/components/Restoration/RestorationType.vue:94]: Actually, it should be initialized to null. Currently you are initializing it to an empty string, but it's not a string.

What I mean is: what if you remove this code from `mounted()` ? Is there any issue with that?

Comment (JazzarKarim) [reply src/components/Restoration/RestorationType.vue:91]: Thing is, in the mounted block, I am auto-selecting either the Full Restoration or Limited Restoration radio buttons when a user loads their previous draft. Without this code, when a user loads a draft, both radio buttons are not selected.

mounted

File: src/components/Restoration/RestorationType.vue
Comment (severinbeauvais) [inline src/components/Restoration/RestorationType.vue:78]: Please add type.

File: src/store/getters/state-getters.ts
Comment (severinbeauvais) [inline src/store/getters/state-getters.ts:611]: Incorrect comment.

File: src/components/Restoration/RestorationType.vue
Comment (severinbeauvais) [reply src/components/Restoration/RestorationType.vue:91]: Doesn't the first radio group get set as soon as the draft restoration filing is loaded?

Comment (severinbeauvais) [reply src/components/Restoration/RestorationType.vue:91]: Does it work correctly atm or does something need to be figured out?

Comment (JazzarKarim) [reply src/components/Restoration/RestorationType.vue:91]: > Doesn't the first radio group get set as soon as the draft restoration filing is loaded?

Without the mounted block, the expansion panel would work, but the radio buttons wouldn't be selected. They would both be empty as in the screenshot above.

File: src/components/Restoration/RestorationType.vue
Comment (severinbeauvais) [inline src/components/Restoration/RestorationType.vue:78]: Isn't it of type RestorationTypes?

**Review by severinbeauvais — APPROVED** [review]

**Comment by severinbeauvais** [comment]
Please get 1 or 2 more approved reviews before merging this.

File: src/components/Restoration/RestorationType.vue
Comment (JazzarKarim) [reply src/components/Restoration/RestorationType.vue:78]: You're right. So sorry about that 😢

**Comment by JazzarKarim** [comment]
> Please get 1 or 2 more approved reviews before merging this.

@jonathan-longe @ketaki-deodhar @ChrisGabel23 Please whenever you can. Thanks a lot!

**Review by ChrisGabel23 — APPROVED** [review]

**Review by ketaki-deodhar — APPROVED** [review]

**Pull request merged**
